polynote-frontend/polynote/messages.js

```diff
@@ -211,6 +211,34 @@ export class CellResult extends Message {
 CellResult.codec = combined(shortStr, int16, Result.codec).to(CellResult);
 
 
+export class NotebookUpdate extends Message {
+
+    static unapply(inst) { return [inst]; }
+
+    /**
+     * Transform a so that it has the same effect when applied after b. Returns transformed a.
+     * @seeScala polynote.messages.NotebookUpdate#rebase
+     */
+    static rebase(a, b) {
+        if (b instanceof Array) {
+            let accum = a;
+            b.forEach(update => {
+                accum = NotebookUpdate.rebase(accum, update);
+            });
+            return accum;
+        }
+
+        if (a instanceof InsertCell && b instanceof InsertCell && a.after === b.after) {
+            return new InsertCell(a.path, a.globalVersion, a.localVersion, b.cell.id);
+        } else if (a instanceof UpdateCell && b instanceof UpdateCell && a.id === b.id) {
+            return new UpdateCell(a.path, a.globalVersion, a.localVersion, ContentEdit.rebaseEdits(a.edits, b.edits));
+        } else {
+            return a;
+        }
```

Comment:
could you use the new `match` here? :)

Reply:
I guess it wouldn't be as nice because you're dealing with two values here.

Reply:
Yeah... and the more I try to implement Scala in JavaScript the more I feel like I probably shoulda just gone with scalajs to begin with... 😞

Reply:
I wonder how hard it would be to switch over... presumably we could start with something small like just replacing messages, say, and slowly migrate the rest? I've never used scalaJS so 🤷‍♂️